src/libasr/pass/replace_symbolic.cpp

```diff
@@ -294,6 +291,51 @@ class ReplaceSymbolicVisitor : public PassUtils::PassVisitor<ReplaceSymbolicVisi
         pass_result.push_back(al, stmt);
     }
 
+    void process_binary_operator(Allocator &al,  const Location &loc, ASR::IntrinsicFunction_t* x, SymbolTable* module_scope,
+        const std::string& new_name, ASR::expr_t* target) {
+            ASR::expr_t* value1 = x->m_args[0];
+            ASR::expr_t* value2 = x->m_args[1];
+
+            if (ASR::is_a<ASR::IntrinsicFunction_t>(*value1) || ASR::is_a<ASR::Cast_t>(*value1)) {
+                if (ASR::is_a<ASR::IntrinsicFunction_t>(*value1)) {
+                    this->visit_IntrinsicFunction(*ASR::down_cast<ASR::IntrinsicFunction_t>(value1));
+                } else {
+                    this->visit_Cast(*ASR::down_cast<ASR::Cast_t>(value1));
+                }
+                ASR::symbol_t* var_sym1 = current_scope->get_symbol(symengine_stack.pop());
+                value1 = ASRUtils::EXPR(ASR::make_Var_t(al, loc, var_sym1));
+            }
```

Comment:
```suggestion
            if (ASR::is_a<ASR::IntrinsicFunction_t>(*value1)) {
                this->visit_IntrinsicFunction(*ASR::down_cast<ASR::IntrinsicFunction_t>(value1));
            } else if (ASR::is_a<ASR::Cast_t>(*value1)) {
                this->visit_Cast(*ASR::down_cast<ASR::Cast_t>(value1));
            } else {
                LCOMPILERS_ASSERT(false);
            }
            ASR::symbol_t* var_sym1 = current_scope->get_symbol(symengine_stack.pop());
            value1 = ASRUtils::EXPR(ASR::make_Var_t(al, loc, var_sym1));
```
Make sense?

Reply:
wait so the logic here goes as follows , `value1` can possibly not be an `IntrinsicFunction` or a `Cast` and in those cases we don't need to edit `value1`. Only if `value1` is either one of those , then we need to edit value1 as done here . 
Hence through my code I just checked if it is one of them ... if yes I edit `value1` through this.
```
            ASR::symbol_t* var_sym1 = current_scope->get_symbol(symengine_stack.pop());
            value1 = ASRUtils::EXPR(ASR::make_Var_t(al, loc, var_sym1));
```
Are these the same ?

Reply:
We have encountered only three cases (`Var_t`, `IntrinsicFunction_t` and `Cast_t`), right?
Any other cases?

Reply:
Also, I think we should only allow CPtr type; for the rest, we have to throw an error or handle it by converting to CPtr type. 

Reply:
> Any other cases?

No , no other case .

Reply:
I am a bit confused about your suggestion here , we only need to create `var_sym` and edit `value` if it is either of these 2 cases `IntrinsicFunction_t` and `Cast_t`. So I am not sure if something like this makes sense 
```
    void process_binary_operator(Allocator &al,  const Location &loc, ASR::IntrinsicFunction_t* x, SymbolTable* module_scope,
        const std::string& new_name, ASR::expr_t* target) {
            ASR::expr_t* value1 = x->m_args[0];
            ASR::expr_t* value2 = x->m_args[1];

            if (ASR::is_a<ASR::IntrinsicFunction_t>(*value1)) {
                this->visit_IntrinsicFunction(*ASR::down_cast<ASR::IntrinsicFunction_t>(value1));
            } else if (ASR::is_a<ASR::Cast_t>(*value1)) {
                this->visit_Cast(*ASR::down_cast<ASR::Cast_t>(value1));
            } else {
                LCOMPILERS_ASSERT(false);
            }
            ASR::symbol_t* var_sym1 = current_scope->get_symbol(symengine_stack.pop());
            value1 = ASRUtils::EXPR(ASR::make_Var_t(al, loc, var_sym1));

            if (ASR::is_a<ASR::IntrinsicFunction_t>(*value2)) {
                this->visit_IntrinsicFunction(*ASR::down_cast<ASR::IntrinsicFunction_t>(value2));
            } else if (ASR::is_a<ASR::Cast_t>(*value2)) {
                this->visit_Cast(*ASR::down_cast<ASR::Cast_t>(value2));
            } else {
                LCOMPILERS_ASSERT(false);
            }
            ASR::symbol_t* var_sym2 = current_scope->get_symbol(symengine_stack.pop());
            value2 = ASRUtils::EXPR(ASR::make_Var_t(al, loc, var_sym2));

        perform_symbolic_binary_operation(al, loc, module_scope, new_name, target, value1, value2);
    }
```

Reply:
> Also, I think we should only allow CPtr type; for the rest, we have to throw an error or handle it by converting to CPtr type.

Also a bit confused about this . Could you show how this would help !

Reply:
The arguments of IntrinisicFunction can be of any type, so we have to make sure we only pass `CPtr` to the `BindC` Function. If this is already handled while creating IntrinisicFunction, then no need to check for the type, I think.

Reply:
Please ignore this review, I realised this: https://github.com/lcompilers/lpython/pull/2263#discussion_r1289655215 and provided a new suggestion: https://github.com/lcompilers/lpython/pull/2263#discussion_r1289625227

Reply:
Actually if you look at the `visit_IntrinsicFunction` now, I think we are handling these things 
```
    void visit_IntrinsicFunction(const ASR::IntrinsicFunction_t &x) {
        if(x.m_type && x.m_type->type == ASR::ttypeType::SymbolicExpression) {
            SymbolTable* module_scope = current_scope->parent;

            ASR::ttype_t *type = ASRUtils::TYPE(ASR::make_SymbolicExpression_t(al, x.base.base.loc));
            std::string symengine_var = symengine_stack.push();
            ASR::symbol_t *arg = ASR::down_cast<ASR::symbol_t>(ASR::make_Variable_t(
                al, x.base.base.loc, current_scope, s2c(al, symengine_var), nullptr, 0, ASR::intentType::Local,
                nullptr, nullptr, ASR::storage_typeType::Default, type, nullptr,
                ASR::abiType::BindC, ASR::Public, ASR::presenceType::Required, false));
            current_scope->add_symbol(s2c(al, symengine_var), arg);
            for (auto &item : current_scope->get_scope()) {
                if (ASR::is_a<ASR::Variable_t>(*item.second)) {
                    ASR::Variable_t *s = ASR::down_cast<ASR::Variable_t>(item.second);
                    this->visit_Variable(*s);
                }
            }

            ASR::IntrinsicFunction_t &xx = const_cast<ASR::IntrinsicFunction_t&>(x);
            ASR::expr_t* target = ASRUtils::EXPR(ASR::make_Var_t(al, x.base.base.loc, arg));
            process_intrinsic_function(al, x.base.base.loc, &xx, module_scope, target);
        }
    }
```

What happens here is firstly we only want to edit those intrinsic function nodes which are of type `SymbolicExpression` . And also here we create a symbolic variables (`stack0` or something from the `symengine_stack`) then run it through `visit_Variable` so that 
1) The 4 statements which are the following are defined for the variable.
```
    _stack0: i64 = i64(0)
    stack0: CPtr = empty_c_void_p()
    p_c_pointer(pointer(_stack0, i64), stack0)
    basic_new_stack(stack0)
```

Reply:
Yup, Let's check the type of `stack0` to be `CPtr`.

Reply:
So I think how this goes is 
1) We only edit an intrinsic function of type SymbolicExpression
2) And while going from AST to ASR there are well defined rules as to what the arguments types can be. Like for binary operators
```
    static inline ASR::asr_t* create_##X(Allocator& al, const Location& loc,               \
            Vec<ASR::expr_t*>& args,                                                       \
            const std::function<void (const std::string &, const Location &)> err) {       \
        if (args.size() != 2) {                                                            \
            err("Intrinsic function `"#X"` accepts exactly 2 arguments", loc);             \
        }                                                                                  \
                                                                                           \
        for (size_t i = 0; i < args.size(); i++) {                                         \
            ASR::ttype_t* argtype = ASRUtils::expr_type(args[i]);                          \
            if(!ASR::is_a<ASR::SymbolicExpression_t>(*argtype)) {                          \
                err("Arguments of `"#X"` function must be of type SymbolicExpression",     \
                args[i]->base.loc);                                                        \
            }                                                                              \
        }  
```
Which means the arguments inside a symbolic type intrinsic function would also be symbolic and visit_IntrinsicFunction should be able to handle them.

Reply:
Sure now that we have a proper framework in `intrinsic_registry.h` using the `create`, `eval` and `instantiate` functions , wherein we actually handle type check through create and also eval through some cases , I haven't disturbed that framework as of now.